hack/check-state.sh

```diff
@@ -0,0 +1,210 @@
+
+        if [ "x$?" == "x0" ]; then
+            break
+        fi
+        if [ "$retry_count_hco" -ge "$MAX_GET_HCO_RETRY" ]; then
+            echo "can't get operator state kind: $kind name: $name. commmand failed repeatedly. $CMD status: $? "
+            exit 1
+        fi
+        sleep 5
+        ((retry_count_hco=retry_count_hco+1))
+    done
+}
+
+
+function get_reconcile_json() {
+    RECONCILE_COMPLETED_JSON=$(echo "$HCO_JSON" | sed -e 's/.*ReconcileComplete: "\([^"]*\)" "\([^"]*\)" "\([^"]*\)".*/Status: \1 Reason: \2 Message: \3/')   
+    RECONCILE_COMPLETED=$(echo "$HCO_JSON" | sed -e 's/.*ReconcileComplete: "\([^"]*\)".*/\1/')   
+    if [ "x$RECONCILE_COMPLETED" != 'xTrue' ] && [ "x$RECONCILE_COMPLETED" != 'xFalse' ]; then
+        echo "Error: ReconcileCompleted not valid: $RECONCILE_COMPLETED_JSON"
+        exit 1
+    fi
+}
+
+function get_application_available_json() {
+    APPLICATION_AVAILABLE_JSON=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationAvailable: "\([^"]*\)" "\([^"]*\)" "\([^"]*\)".*/Status: \1 Reason: \2 Message: \3/')   
+    APPLICATION_AVAILABLE=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationAvailable: "\([^"]*\)".*/\1/')   
+    if [ "x$APPLICATION_AVAILABLE" != 'xTrue' ] && [ "x$APPLICATION_AVAILABLE" != 'xFalse' ]; then
+        echo "Error: ApplicationAvailable not valid: $APPLICATION_AVAILABLE_JSON"
+        exit 1
+    fi
+}
+
+function get_operation_progressing_json() {
+    OPERATION_PROGRESSING_JSON=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationProgressing: "\([^"]*\)" "\([^"]*\)" "\([^"]*\)".*/Status: \1 Reason: \2 Message: \3/')   
+    OPERATION_PROGRESSING=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationProgressing: "\([^"]*\)".*/\1/')   
+     
+    if [ "x$OPERATION_PROGRESSING" != 'xTrue' ] && [ "x$OPERATION_PROGRESSING" != 'xFalse' ]; then
+        echo "Error: OperationProgressing not valid: $OPERATION_PROGRESSING_JSON"
+        exit 1
+    fi
+}
+
+function get_application_degraded_json() {
+    APPLICATION_DEGRADED_JSON=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationDegraded: "\([^"]*\)" "\([^"]*\)" "\([^"]*\)".*/Status: \1 Reason: \2 Message: \3/')   
+    APPLICATION_DEGRADED=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationDegraded: "\([^"]*\)".*/\1/')   
+  
+    if [ "x$APPLICATION_DEGRADED" != 'xTrue' ] && [ "x$APPLICATION_DEGRADED" != 'xFalse' ]; then
+        echo "Error: ApplicationDegraded not valid: $APPLICATION_DEGRADED_JSON"
+        exit 1
+    fi
+}
+
+
+function check_operator_up() {
+   local kind=$1
+   local name=$2
+   local namespace=$3
+
+cat <<EOF
+
+Checking operator kind: $kind name: $name 
+
+EOF
+
+   retry_count=0
+   while [ "$retry_count" -lt "$MAX_APP_AVAILABLE_CHECKS" ]; do
+      get_operator_json $kind $name $namespace 
```

Comment:
I think this script would benefit from eliminating some of the functions (I personally think all of the `get_*_json` functions can go away), and explicitly getting the condition you want with something like:

```
${CMD} get $kind $name $namespace -o go-template='{{range .status.conditions }}{{ if eq .type "Available" }}{{ .status }}{{ end }}'
```

You only really need the `reason` and `message fields to tell us humans whats broken and why...in which case you would probably want to just do a `kubectl describe` on the object so we can see the whole thing in a glance.

Reply:
I will remove the get_*_json functions; However I don't think we want to have call to kubectl/oc get for every condition check, that's why I have one kubectl get to retrieve all the data and then parse it for each condition (previously I got the json for the whole condition and extracted each field by means of jq - however jq is not installed on all platforms so I switched it to go-template)

Reply:
> However I don't think we want to have call to kubectl/oc get for every condition check (these take a long time and things may change between runs)

My guess is that this is because of the performance hit of using `kubectl` to talk to an api server to grab the conditions that we already have? If that is the case, my argument would be that since this is a test and not an application itself + the bottleneck is in how long the upgrade takes not these commands keeping up that the improved readability is worth the performance impact given the mitigating circumstances.

Reply:
I am more concerned that the state of the operator may change in between runs of kubectl, so that the script would make an incorrect decision based on old data obtained from a previous run; I think it is therefore reasonable to get the state of all conditions in one run of kubectl.

Reply:
While I'm not totally on board with the concern, I think what you have is fairly easy to reason about so I'm good with it.